az-flow-trigger-dependency-type/kafka-event-trigger/src/main/Java/trigger.kafka/KafkaDepInstanceCollection.java

```diff
@@ -0,0 +1,158 @@
+ * the License.
+ */
+
+package trigger.kafka;
+
+import com.google.common.base.Joiner;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A KafkaDepInstanceCollection that maintain map for topic and event with all the util method.
+ * Enable sharing on regrex match rules, so as to be more efficient.
+ * Structure looks like:
+ * {-Topic1:{
+ * ----Rule1
+ * ---------[List of dependencies]
+ * ----Rule2
+ * ---------[List of dependencies]
+ * }
+ * -Topic2:{
+ * ----Rule1
+ * ---------[List of dependencies]
+ * ----Rule2
+ * ---------[List of dependencies]
+ * }
+ * }
+ *
+ */
+public class KafkaDepInstanceCollection {
+
+  private final static Logger log = LoggerFactory.getLogger(KafkaDepInstanceCollection.class);
+  private final Map<String, Map<String, List<KafkaDependencyInstanceContext>>> topicEventMap;
+
+  public KafkaDepInstanceCollection() {
+    this.topicEventMap = new HashMap<>();
+  }
+
+  public synchronized void add(final KafkaDependencyInstanceContext dep) {
+    final String topic = dep.getTopicName();
+    Map<String, List<KafkaDependencyInstanceContext>> eventMap = this.topicEventMap.get(topic);
+    List<KafkaDependencyInstanceContext> depList;
+    if (eventMap == null) {
+      eventMap = new HashMap<>();
+      depList = new LinkedList<>();
+    } else {
+      depList = eventMap.get(dep.getRegexMatch());
+      if (depList == null) {
+        depList = new LinkedList<>();
+      }
+    }
+    depList.add(dep);
+    eventMap.put(dep.getRegexMatch(), depList);
+    this.topicEventMap.put(topic, eventMap);
+  }
+
+  public boolean hasTopic(final String topic) {
+    return !(this.topicEventMap.get(topic) == null);
+  }
+
+  public List<String> getTopicList() {
+    final List<String> res = new ArrayList<>(this.topicEventMap.keySet());
+    return res;
+  }
+
+  public Set<String> eventInTopic(final String topic, final RegexKafkaDependencyMatcher matcher,
+      final String payload) {
+    final Set<String> res = new HashSet<>();
+    final Map<String, List<KafkaDependencyInstanceContext>> eventMap = this.topicEventMap.get(topic);
+    if (eventMap == null) {
+      return null;
+    }
+    for (final Map.Entry<String, List<KafkaDependencyInstanceContext>> entry : eventMap.entrySet()) {
+      if (matcher.isMatch(payload, entry.getKey())) {
+        res.add(entry.getKey());
+      }
+    }
+    return res;
+  }
+
+  public synchronized List<KafkaDependencyInstanceContext> getDepsByTopicAndEvent(final String topic,
+      final String event) {
+    final Map<String, List<KafkaDependencyInstanceContext>> eventMap = this.topicEventMap.get(topic);
+    if (eventMap != null) {
+      return eventMap.get(event);
+    }
+    return null;
```

Comment:
please return empty set instead of null to avoid null pointer check on the caller side.